doc/feature_derivation.md

```diff
@@ -0,0 +1,85 @@
+# Design: Feature Derivation
+
+This file discusses the details and implementations of "Feature Derivation".
+Please refer to [this](https://medium.com/@SQLFlow/feature-derivation-the-conversion-from-sql-data-to-tensors-833519db1467) blog to
+get the background information before going on.
+
+Target: We need to know the below two information for each column after the
+feature type infer routine:
+
+1. How to transform the column data to tensors, including `tf.SparseTensor`.
+2. What type of feature column should adapt to the column and the parameters for the feature column call.
+
+## Simpler COLUMN Clause
+
+When we have a training table contains many columns that should be used
+for training like https://www.kaggle.com/mlg-ulb/creditcardfraud, it's
+not friendly if we must provide all column names in `COLUMN` clause, we'd
+like to just use `COLUMN *` here.
+
+But, for cases when the columns are of quite different data format, like:
+
+```
+data type: (float, float, float, float, csv_string_for_dense_tensor, csv_string_for_sparse_tensor, int)
+column name: (a , b , c , d , e , f, label)
+```
+
+`COLUMN *` and plus the data in the table may not provide enough information we need:
+
+- whether the `csv_string_for_dense_tensor` should going through an embedding layer
+- what the original "dense shape" for the `csv_string_for_sparse_tensor` column to parse to a `SparseTensor`.
+
+So we need SQLFlow users to provide such information in the SQL statements to make the training
+network specific. We support the below SQL statements for the above case:
+
+```
+SELECT * FROM training_table
+TRAIN DNNClassifier
+WITH someattr=somevalue
+COLUMN *,EMBEDDING(e, 128, "sum"),EMBEDDING(SPARSE(f, [1000000]), 512, "sum")
```

Comment:
@typhoonzero Thanks for this syntax design.

Looks like the `*` in `select * from train_table` and the `*` in `COLUMN *` have different meanings. The first is the shorthand for all fields. And the later is rather complex, whose meaning depends on many other things.

1. Consider `select *, a from train_table`, the result columns are all fields in `train_table` plus a duplicated `a`. However, `COLUMN *, a` has the same columns as `COLUMN *`. The meaning of `*` is depending on `a`.
1. The `COLUMN *` in `COLUMN * LABEL label` doesn't include the label field. The meaning of `*` is depending on the label clause.

How about introducing a function named `MATCH(regexp)` that will expand the regexp to match column names. So we can write `MATCH("[a-d]")` to represent columns `a`, `b`, `c` and `d`. So the train SQL would become
```SQL
SELECT * FROM training_table
TRAIN DNNClassifier
WITH someattr=somevalue
COLUMN MATCH("[a-d]"),EMBEDDING(e, 128, "sum"),EMBEDDING(SPARSE(f, [1000000]), 512, "sum")
LABEL label
INTO my_model_file;
```
In the case of [credit fraud](https://www.kaggle.com/mlg-ulb/creditcardfraud) dataset, we want to select `V1`, `V2`, ..., `V19`, we can write `MATCH("V\d+")`, which would exclude field `time` and `class`.

Reply:
As we discussed this morning, I am sure that what must be done by the SQL statements are:

1. Describe what columns to read from. It could be all columns in the table or several columns in some pattern.
1. Describe what are the columns used for training samples ( or input X ), and what is the columns used for the label (Y).
1. If there are a lot of X columns, we should have a fast way to match them all, e.g. credit card fraud dataset have time, v1~v18, amount features, we can match them by `time, MATCH(v[0-9][0-9]), amount`.
1. Describe how to parse the column data into tensors, e.g. parse a SCV string to a `tf.SparseTensor` or `tf.Tensor`, or preprocess the tensors like image random crop etc.
1. Describe how the feature column is assembled to the network, like whether to use embedding for the column, should the embedding column use shared weights etc.